WordPress/src/test/java/org/wordpress/android/ui/uploads/UploadActionUseCaseTest.kt

```diff
@@ -0,0 +1,508 @@
+package org.wordpress.android.ui.uploads
+
+import androidx.arch.core.executor.testing.InstantTaskExecutorRule
+import com.nhaarman.mockitokotlin2.any
+import com.nhaarman.mockitokotlin2.mock
+import com.nhaarman.mockitokotlin2.whenever
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.junit.MockitoJUnitRunner
+import org.wordpress.android.fluxc.model.PostModel
+import org.wordpress.android.fluxc.model.SiteModel
+import org.wordpress.android.fluxc.model.post.PostStatus
+import org.wordpress.android.fluxc.store.UploadStore
+import org.wordpress.android.ui.posts.PostUtilsWrapper
+import org.wordpress.android.ui.uploads.UploadActionUseCase.UploadAction
+import org.wordpress.android.util.DateTimeUtils
+import java.util.Date
+
+private val POST_STATE_DRAFT = PostStatus.DRAFT.toString()
+
+@RunWith(MockitoJUnitRunner::class)
+class UploadActionUseCaseTest {
+    @get:Rule val rule = InstantTaskExecutorRule()
+
+    @Test
+    fun `uploadAction is UPLOAD when changes confirmed`() {
+        val uploadActionUseCase = createUploadActionUseCase()
+
+        val post = createPostModel(changesConfirmed = true)
+        // Act
+        val action = uploadActionUseCase.getUploadAction(post)
+
+        // Assert
+        assertThat(action).isEqualTo(UploadAction.UPLOAD)
+    }
+
+    @Test
+    fun `uploadAction is UPLOAD_AS_DRAFT when changes not confirmed and is local draft`() {
+        val uploadActionUseCase = createUploadActionUseCase()
+
+        val post = createPostModel(changesConfirmed = false, isLocalDraft = true)
+        // Act
+        val action = uploadActionUseCase.getUploadAction(post)
+
+        // Assert
+        assertThat(action).isEqualTo(UploadAction.UPLOAD_AS_DRAFT)
+    }
+
+    @Test
+    fun `uploadAction is REMOTE_AUTO_SAVE when changes not confirmed and isn't local draft`() {
+        val uploadActionUseCase = createUploadActionUseCase()
+
+        val post = createPostModel(changesConfirmed = false, isLocalDraft = false)
+        // Act
+        val action = uploadActionUseCase.getUploadAction(post)
+
+        // Assert
+        assertThat(action).isEqualTo(UploadAction.REMOTE_AUTO_SAVE)
+    }
+
+    @Test
+    fun `autoUploadAction is DO NOTHING when the post is older than 2 days`() {
```

Comment:
I had expected that we would sill upload local drafts but I guess `DO_NOTHING` is safer. In any case, I think we should cover permutations that we consider as important. For this test, what do you think of this?

```kt
    @Test
    fun `autoUploadAction is DO NOTHING when the post is older than 2 days`() {
        // Arrange
        val uploadActionUseCase = createUploadActionUseCase()

        val twoDaysAgoTimestamp = run {
            val twoDaysInSeconds = 60 * 60 * 24 * 2
            val twoDaysAgo = (Date().time / 1000) - twoDaysInSeconds
            DateTimeUtils.iso8601FromTimestamp(twoDaysAgo)
        }

        val posts = listOf(
                createPostModel(dateLocallyChanged = twoDaysAgoTimestamp),
                createPostModel(dateLocallyChanged = twoDaysAgoTimestamp, isLocalDraft = true, changesConfirmed = true),
                createPostModel(dateLocallyChanged = twoDaysAgoTimestamp, isLocalDraft = false, changesConfirmed = true)
        )

        val siteModel: SiteModel = createSiteModel()

        // Act and Assert
        assertThat(posts).allSatisfy { post ->
            val action = uploadActionUseCase.getAutoUploadAction(post, siteModel)

            assertThat(action).isEqualTo(UploadAction.DO_NOTHING)
        }
    }
```

Reply:
>I had expected that we would sill upload local drafts

My understanding was that one of the main reasons why we want to have this "2 day" guard is that if the upload doesn't work and the user decides to update the post on the web, we don't want to override the remote post after that. If we uploaded it as a draft we might "un-publish" a post. So as you said I think it's safer to simply ignore it.

>I think we should cover permutations that we consider as important. For this test, what do you think of this?

Tbh I'm not sure about it. Can you elaborate why you think it's important to test the permutations just for this specific scenario and not for all the others (especially since the 2 day guard is just additional guard which isn't critical at all)? I mean ideally we'd cover 100% of all the possible scenarios, but I don't think it's the goal of our unit tests. I agree that you could change the `getAutoUploadAction` so the tests would still succeed but the logic would be broken, but you would need to put an effort into it - eg. add conditions like `if(isOlderThan2Days && !isLocalDraft)` or similar. 
Either case I added the new test as it can't hurt us to have more unit tests. I'd still like to hear your opinion;). Thanks

Reply:
Thank you, @malinajirka. 

I don't think it's important to cover all scenarios. Just the important ones. And, for me, it's important to cover the scenario that “local drafts will also not be automatically uploaded if they're 2 days old”. This felt important to me because when reviewing this, as I mentioned, I had expected that local drafts will **always** be automatically uploaded. But apparently, we don't. 

And it's fine that we don't auto-upload, I just requested that it will be added in the test to _document_ that this is how we intended it to work. 🙂 

Please let me know if I didn't explain this well. 😅

Reply:
Makes sense;). Thanks for the clarification!